### DISCUSSION TITLE
new visitor with scope info

### DISCUSSION THREAD

**Discussion by JSerFeng** [swc-project/swc#6843] [Ideas]
Scope is hard to deal with in some certain plugi transform. It is very useful if we can get scope information of current visiting node.
We dont need to use Arc or Rc, we can make a new Scope instance every time we enter a node, like what have done in visit_with_path.
We can do this by:
- In resolver, add a stxt in block or function or any other node that will need to create a new scope. And then, every time we enter a node that need to create scope, we can iterate all nodes with the same stxt at current scope, then create scope instance based on that, and pass it to the visitor.

How do you think of that, willing to contribute.

Also if you think this is a good idea, what field should the scope contains, and what should I do if I want to both use scope and path feature, visit_scope_and_path? Or add visit with context, and the context can have the field path and scope based on the feature.

**Comment by kdy1** [comment]
Adding the contexts to `BlockStmt`/`Function` from the resolver pass is fine, but I'm not sure about the extra.
I think it can be a third-party crate at the start, though.

**Reply by JSerFeng** [reply]
Thanks for reply, that's great enough, I'll make a PR to add stxt to `BlockStmt` and `Function`

**Comment by hyf0** [comment]
@JSerFeng You could use https://github.com/swc-project/swc/tree/main/crates/swc_ecma_usage_analyzer.

The minifier use it to do scope analysis. It might be enough/overload for simple use cases.

**Reply by JSerFeng** [reply]
But that does not have information about the relevance of scopes.

I want this feature mainly because I want to know a variable is from which scope beyond current scope.

```javascript
function A() {
  const a = 1;

  {
    const b = 2;

    return () => {
      return a + b // We are here
    }
  }
}
```

I want to know `a` and `b` inside `console.log(a, b)` are from which scope, and the relationship between them.

so I can transform that into

```javascript
function A() {
  const a = 1;

  {
    const b = 2;

    let _res; // hoist the result to the closest accessed scope

    return () => {
      return _res || (_res = a + b)
    }
  }
}
```

It can be used to port babel-plugin-react-const-element

**Reply by kdy1** [reply]
Actually, the usage analyzer stores scope info based on the `SyntaxContext`.

https://rustdoc.swc.rs/swc_ecma_usage_analyzer/analyzer/storage/trait.Storage.html#tymethod.scope


But then the problem is that you need to apply https://rustdoc.swc.rs/swc_ecma_minifier/fn.unique_scope.html too.
I think we should move it to the usage analyzer.

**Reply by kdy1** [reply]
I decided not to apply syntax contexts to the blocks or functions from the `resolver` because there are lots of passes between the 
resolver and the minfiier, and there's no constraint about the syntax contexts of those nodes.

**Reply by JSerFeng** [reply]
Got it

**Reply by JSerFeng** [reply]
Is there any useful tools I can use to deal with something about scope ?

**Reply by kdy1** [reply]
AFAIK, no. The visitors I mentioned were enough for me.

**Reply by JSerFeng** [reply]
Alright, thanks for reply
